File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        /*
        1. Create a copy forwards and a copy backwards or reversed
        2. Iterate repeatedly alternating forward and backward into merged list
        3. There may be one element left if odd. Handle it
        bckwd = new node
        curr = head
        prev = None
        while curr and curr.next:
            next = curr.next
            curr.next = prev
         */
        if (!head) return;

        vector<ListNode*> nodes;
        ListNode* cur = head;
        while(cur){
            nodes.push_back(cur);
            cur = cur->next;
        }
        int i = 0;
        int j = nodes.size() - 1;
        while (i < j){
            nodes[i++]->next = nodes[j];
            if (i >= j) break;
            nodes[j--]->next = nodes[i];
        }
        nodes[i]->next= nullptr;
    }
};

File: 0143-reorder-list/0143-reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        // 1. Find the midpoint of the LL
        ListNode slow = head;
        ListNode fast = head.next;
        while (fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
        }
        // 2. reverse the LL from the midpoint on
        ListNode curr = slow.next;
        ListNode prev = null;
        slow.next = null;
        while (curr != null){
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }

        // 3. Merge the two linked lists
        ListNode second = prev;
        ListNode first = head;
        while (first != null && second != null){
            ListNode temp1 = first.next;
            ListNode temp2 = second.next;
            first.next = second;
            second.next = temp1;
            first = temp1;
            second = temp2;
        }
        
    }
}